File: sos-app/apps/mobile-ios/SOSApp/Networking/APIClient.swift
import Foundation
import Alamofire

class APIClient {
    static let shared = APIClient()
    private let baseURL = "http://localhost:3000/api/v1"
    private var accessToken: String?

    private init() {}

    // MARK: - Authentication
    func login(email: String, password: String, completion: @escaping (Result<AuthResponse, Error>) -> Void) {
        let parameters: [String: Any] = [
            "email": email,
            "password": password
        ]

        AF.request("\(baseURL)/auth/login",
                   method: .post,
                   parameters: parameters,
                   encoding: JSONEncoding.default)
            .validate()
            .responseDecodable(of: AuthResponse.self) { response in
                switch response.result {
                case .success(let authResponse):
                    self.accessToken = authResponse.tokens.accessToken
                    KeychainHelper.save(token: authResponse.tokens.accessToken, key: "accessToken")
                    KeychainHelper.save(token: authResponse.tokens.refreshToken, key: "refreshToken")
                    completion(.success(authResponse))
                case .failure(let error):
                    completion(.failure(error))
                }
            }
    }

    func register(userData: RegisterData, completion: @escaping (Result<AuthResponse, Error>) -> Void) {
        AF.request("\(baseURL)/auth/register",
                   method: .post,
                   parameters: userData,
                   encoder: JSONParameterEncoder.default)
            .validate()
            .responseDecodable(of: AuthResponse.self) { response in
                switch response.result {
                case .success(let authResponse):
                    self.accessToken = authResponse.tokens.accessToken
                    KeychainHelper.save(token: authResponse.tokens.accessToken, key: "accessToken")
                    completion(.success(authResponse))
                case .failure(let error):
                    completion(.failure(error))
                }
            }
    }

    // MARK: - Emergency
    func triggerEmergency(type: EmergencyType, description: String?, location: LocationData?, completion: @escaping (Result<Emergency, Error>) -> Void) {
        guard let token = accessToken ?? KeychainHelper.load(key: "accessToken") else {
            completion(.failure(APIError.unauthorized))
            return
        }

        var parameters: [String: Any] = ["type": type.rawValue]
        if let desc = description { parameters["description"] = desc }
        if let loc = location {
            parameters["location"] = [
                "latitude": loc.latitude,
                "longitude": loc.longitude
            ]
        }

        let headers: HTTPHeaders = ["Authorization": "Bearer \(token)"]

        AF.request("\(baseURL)/emergency/trigger",
                   method: .post,
                   parameters: parameters,
                   encoding: JSONEncoding.default,
                   headers: headers)
            .validate()
            .responseDecodable(of: EmergencyResponse.self) { response in
                switch response.result {
                case .success(let emergencyResponse):
                    completion(.success(emergencyResponse.emergency))
                case .failure(let error):
                    completion(.failure(error))
                }
            }
    }

    func resolveEmergency(emergencyId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let token = accessToken ?? KeychainHelper.load(key: "accessToken") else {
            completion(.failure(APIError.unauthorized))
            return
        }

        let headers: HTTPHeaders = ["Authorization": "Bearer \(token)"]

        AF.request("\(baseURL)/emergency/\(emergencyId)/resolve",
                   method: .post,
                   headers: headers)
            .validate()
            .response { response in
                if let error = response.error {
                    completion(.failure(error))
                } else {
                    completion(.success(()))
                }
            }
    }

    // MARK: - Contacts
    func getContacts(completion: @escaping (Result<[EmergencyContact], Error>) -> Void) {
        guard let token = accessToken ?? KeychainHelper.load(key: "accessToken") else {
            completion(.failure(APIError.unauthorized))
            return
        }

        let headers: HTTPHeaders = ["Authorization": "Bearer \(token)"]

        AF.request("\(baseURL)/contacts",
                   method: .get,
                   headers: headers)
            .validate()
            .responseDecodable(of: ContactsResponse.self) { response in
                switch response.result {
                case .success(let contactsResponse):
                    completion(.success(contactsResponse.contacts))
                case .failure(let error):
                    completion(.failure(error))
                }
            }
    }
}

enum APIError: Error {
    case unauthorized
    case invalidResponse
}

File: sos-app/apps/web/src/lib/api-client.ts
import axios, { AxiosInstance, AxiosError } from 'axios';
import Cookies from 'js-cookie';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:3000';

class APIClient {
  private client: AxiosInstance;

  constructor() {
    this.client = axios.create({
      baseURL: API_URL,
      headers: {
        'Content-Type': 'application/json',
      },
      withCredentials: true,
    });

    // Request interceptor to add auth token
    this.client.interceptors.request.use(
      (config) => {
        const token = Cookies.get('auth_token');
        if (token) {
          config.headers.Authorization = `Bearer ${token}`;
        }
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );

    // Response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      async (error: AxiosError) => {
        if (error.response?.status === 401) {
          // Token expired or invalid
          Cookies.remove('auth_token');
          Cookies.remove('refresh_token');
          if (typeof window !== 'undefined') {
            window.location.href = '/login';
          }
        }
        return Promise.reject(error);
      }
    );
  }

  // Authentication endpoints
  async login(email: string, password: string) {
    const response = await this.client.post('/api/v1/auth/login', {
      email,
      password,
    });
    return response.data;
  }

  async register(data: {
    email: string;
    password: string;
    firstName: string;
    lastName: string;
    phoneNumber: string;
  }) {
    const response = await this.client.post('/api/v1/auth/register', data);
    return response.data;
  }

  async logout() {
    const response = await this.client.post('/api/v1/auth/logout');
    Cookies.remove('auth_token');
    Cookies.remove('refresh_token');
    return response.data;
  }

  async refreshToken() {
    const refreshToken = Cookies.get('refresh_token');
    const response = await this.client.post('/api/v1/auth/refresh', {
      refreshToken,
    });
    return response.data;
  }

  // User endpoints
  async getProfile() {
    const response = await this.client.get('/api/v1/users/profile');
    return response.data;
  }

  async updateProfile(data: Partial<any>) {
    const response = await this.client.put('/api/v1/users/profile', data);
    return response.data;
  }

  // Emergency endpoints
  async triggerEmergency(data: {
    type: string;
    description?: string;
    location?: { latitude: number; longitude: number };
  }) {
    const response = await this.client.post('/api/v1/emergency/trigger', data);
    return response.data;
  }

  async cancelEmergency(emergencyId: string) {
    const response = await this.client.post(
      `/api/v1/emergency/${emergencyId}/cancel`
    );
    return response.data;
  }

  async resolveEmergency(emergencyId: string) {
    const response = await this.client.post(
      `/api/v1/emergency/${emergencyId}/resolve`
    );
    return response.data;
  }

  async getActiveEmergency() {
    const response = await this.client.get('/api/v1/emergency/active');
    return response.data;
  }

  async getEmergencyHistory(params?: {
    page?: number;
    limit?: number;
    startDate?: string;
    endDate?: string;
  }) {
    const response = await this.client.get('/api/v1/emergency/history', {
      params,
    });
    return response.data;
  }

  async getEmergencyById(emergencyId: string) {
    const response = await this.client.get(`/api/v1/emergency/${emergencyId}`);
    return response.data;
  }

  // Contact endpoints
  async getContacts() {
    const response = await this.client.get('/api/v1/contacts');
    return response.data;
  }

  async addContact(data: {
    name: string;
    phoneNumber: string;
    email?: string;
    relationship: string;
    priority: number;
  }) {
    const response = await this.client.post('/api/v1/contacts', data);
    return response.data;
  }

  async updateContact(contactId: string, data: Partial<any>) {
    const response = await this.client.put(
      `/api/v1/contacts/${contactId}`,
      data
    );
    return response.data;
  }

  async deleteContact(contactId: string) {
    const response = await this.client.delete(`/api/v1/contacts/${contactId}`);
    return response.data;
  }

  // Medical profile endpoints
  async getMedicalProfile() {
    const response = await this.client.get('/api/v1/medical/profile');
    return response.data;
  }

  async updateMedicalProfile(data: {
    bloodType?: string;
    allergies?: string[];
    medications?: Array<{ name: string; dosage: string }>;
    conditions?: string[];
  }) {
    const response = await this.client.put('/api/v1/medical/profile', data);
    return response.data;
  }

  // Location endpoints
  async updateLocation(emergencyId: string, location: {
    latitude: number;
    longitude: number;
    accuracy?: number;
  }) {
    const response = await this.client.post('/api/v1/location/update', {
      emergencyId,
      ...location,
    });
    return response.data;
  }

  async getLocationTrail(emergencyId: string) {
    const response = await this.client.get(
      `/api/v1/location/trail/${emergencyId}`
    );
    return response.data;
  }
}

export const apiClient = new APIClient();
